File: Greedy/boj_1931_회의실 배정.cpp
//https://www.acmicpc.net/problem/1931
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

bool cmp(const pair<int,int>a, const pair<int,int> b) {
	if (a.second == b.second)return a.first < b.first;
	else return a.second < b.second;
}

int main() {
	ios::sync_with_stdio(false);
	cin.tie(0);
	cout.tie(0);

	int n;
	int answer = 0;
	cin >> n;
	vector<pair<int, int>> vt(n);
	int end = 0;
	for (int i = 0; i < n; i++) {
		int a, b;
		cin >> a >> b;
		vt[i] = { a, b };
	}
	sort(vt.begin(), vt.end(), cmp);
	
	for (int i = 0; i < n; i++) {
		if (vt[i].first >= end) {
			answer++;
			end = vt[i].second;
		}
	}
	cout << answer;
	return 0;
}

File: Greedy/boj_1931_회의실 배정.kt
//https://www.acmicpc.net/problem/1931
import java.util.*
val br = System.`in`.bufferedReader()

fun main() = with(System.out.bufferedWriter()){

    //input
    val n = br.readLine().toInt()
    //end시간 기준 오름차순, 같으면 start시간 기준 오름차순
    val pq = PriorityQueue<Pair<Int,Int>>{a,b -> when{
        a.second < b.second -> -1
        a.second == b.second -> {
            when{
                a.first < b.first -> -1
                else -> 1
            }
        }
        else -> 1
    }}

    repeat(n){
        val (from, to) = br.readLine().split(' ').map{it.toInt()}
        pq.add(Pair(from,to))
    }

    var cur=0
    var answer=0
    while(pq.isNotEmpty()){
        val (start,end) = pq.poll()
        if(start<cur)continue
        answer++
        cur=end
    }
    write("$answer")

    close()
}
